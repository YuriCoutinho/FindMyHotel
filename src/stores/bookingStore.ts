import { defineStore } from 'pinia'

export const useBookingStore = defineStore('bookingStore', {
  state: () => ({
    selectedHotel: null as any | null,
  }),
  getters: {
    getSelectedHotel: (state) => state.selectedHotel,
  },
  actions: {
    setSelectedHotel(hotel: any) {
      this.selectedHotel = hotel
    }
  }
})
