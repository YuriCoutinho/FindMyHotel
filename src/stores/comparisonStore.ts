import { defineStore } from 'pinia'

export const useComparisonStore = defineStore('comparisonStore', {
  state: () => ({
    hotelsToCompare: [] as any[]
  }),
  actions: {
    addHotel(hotel: any) {
      if (!this.hotelsToCompare.find((h) => h.id === hotel.id)) {
        this.hotelsToCompare.push(hotel)
      }
    },
    removeHotel(hotelId: string) {
      this.hotelsToCompare = this.hotelsToCompare.filter((hotel) => hotel.id !== hotelId)
    }
  }
})
