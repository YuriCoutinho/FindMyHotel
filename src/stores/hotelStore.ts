import { defineStore } from 'pinia'
import axios from 'axios'
import { hotels as initialHotels } from '@/utils/constants'

export const useHotelStore = defineStore('hotelStore', {
  state: () => ({
    selectedState: null as { name: string; abbreviation: string } | null,
    selectedCity: '' as string | null,
    cities: [] as string[],
    checkInDate: '' as string | null,
    checkOutDate: '' as string | null,
    guests: 1,
    rooms: 1,
    searchIsLoading: false as boolean,
    hotels: [] as any[],
    originalHotels: [] as any[]
  }),
  getters: {
    getSelectedState: (state) => state.selectedState,
    getSelectedCity: (state) => state.selectedCity,
    getCheckInDate: (state) => state.checkInDate,
    getCheckOutDate: (state) => state.checkOutDate,
    getGuests: (state) => state.guests,
    getRooms: (state) => state.rooms,
    getSearchIsLoading: (state) => state.searchIsLoading,
    getHotels: (state) => state.hotels,
    getCities: (state) => state.cities,
    isCitySelectionEnabled: (state) => state.cities.length > 0,
    isFormValid: (state) => {
      return (
        state.selectedState !== null &&
        state.selectedCity !== '' &&
        state.checkInDate !== '' &&
        state.checkOutDate !== '' &&
        state.guests > 0 &&
        state.rooms > 0
      )
    }
  },
  actions: {
    setState(state: { name: string; abbreviation: string } | null) {
      this.selectedState = state
      if (state) {
        this.searchCities(state.abbreviation)
      }
    },
    setCity(city: string | null) {
      this.selectedCity = city
    },
    setCheckInDate(checkInDate: string | null) {
      this.checkInDate = checkInDate
    },
    setCheckOutDate(checkOutDate: string | null) {
      this.checkOutDate = checkOutDate
    },
    setGuests(selectedGuests: number) {
      this.guests = selectedGuests
    },
    setRooms(selectedRooms: number) {
      this.rooms = selectedRooms
    },
    setSearchIsLoading(searchIsLoading: boolean) {
      this.searchIsLoading = searchIsLoading
    },
    setHotels(hotels: any[]) {
      this.hotels = hotels
    },
    sortHotelsByPriceAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.price - b.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByPriceDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.price - a.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.stars - b.stars)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.stars - a.stars)
      this.setHotels(sortedHotels)
    },
    resetHotels() {
      this.setHotels([...this.originalHotels])
    },
    async searchCities(stateAbbreviation: string) {
      this.selectedCity = null
      this.cities = []
      if (stateAbbreviation) {
        try {
          const response = await axios.get(
            `https://servicodados.ibge.gov.br/api/v1/localidades/estados/${stateAbbreviation}/municipios`
          )
          this.cities = response.data.map((municipality: any) => municipality.nome)
        } catch (error) {
          console.error('Error fetching municipalities:', error)
          this.cities = []
        }
      }
    },
    async searchHotels() {
      this.setSearchIsLoading(true)
      try {
        setTimeout(() => {
          this.hotels = [...initialHotels]
          this.originalHotels = [...initialHotels]
          this.setSearchIsLoading(false)
        }, 3000)
      } catch (error) {
        console.error('Error searching hotels:', error)
        this.setSearchIsLoading(false)
      }
    }
  }
})
