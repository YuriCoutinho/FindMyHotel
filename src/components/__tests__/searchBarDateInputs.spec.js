import { mount } from '@vue/test-utils'
import { describe, it, expect, beforeEach } from 'vitest'
import { createPinia, setActivePinia } from 'pinia'
import PrimeVue from 'primevue/config'
import InputMask from 'primevue/inputmask'
import DateInputs from '@/components/SearchBarDateInputs.vue'
import { useHotelStore } from '@/stores/hotelStore'

describe('DateInputs.vue', () => {
  let hotelStore

  beforeEach(() => {
    const pinia = createPinia()
    setActivePinia(pinia)
    hotelStore = useHotelStore()
  })

  const createWrapper = () => {
    return mount(DateInputs, {
      global: {
        plugins: [PrimeVue],
        components: { InputMask }
      }
    })
  }

  it('Should render the component', () => {
    const wrapper = createWrapper()
    expect(wrapper.exists()).toBe(true)
  })

  it('Should initialize with the correct values', () => {
    const wrapper = createWrapper()
    const checkInComponent = wrapper.findAllComponents(InputMask)[0]
    const checkOutComponent = wrapper.findAllComponents(InputMask)[1]

    expect(checkInComponent.props('modelValue')).toBe(hotelStore.getCheckInDate || '')
    expect(checkOutComponent.props('modelValue')).toBe(hotelStore.getCheckOutDate || '')
  })

  it('Should match the snapshot', () => {
    const wrapper = createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Should update the store when check-in date is changed', async () => {
    const wrapper = createWrapper()
    const checkInComponent = wrapper.findAllComponents(InputMask)[0]

    await checkInComponent.setValue('12/12/2024')
    await wrapper.vm.$nextTick()

    expect(hotelStore.getCheckInDate).toBe('12/12/2024')
  })

  it('Should update the store when check-out date is changed', async () => {
    const wrapper = createWrapper()
    const checkOutComponent = wrapper.findAllComponents(InputMask)[1]

    await checkOutComponent.setValue('13/12/2024')
    await wrapper.vm.$nextTick()

    expect(hotelStore.getCheckOutDate).toBe('13/12/2024')
  })
})
