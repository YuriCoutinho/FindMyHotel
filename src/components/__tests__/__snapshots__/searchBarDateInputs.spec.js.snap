// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateInputs.vue > Should match the snapshot 1`] = `
"<div class="flex flex-row items-center gap-4 justify-between md:justify-start"><label for="CheckIn">CheckIn</label><input type="text" class="p-inputtext p-component p-inputmask w-32 text-gray-700 font-medium focus:outline-none focus:ring-1 border rounded-lg px-3 py-2 focus:ring-emerald-500" placeholder="dd/mm/yyyy" data-pc-name="pcinputtext" data-pc-extend="inputtext" data-pc-section="root" value=""></div>
<div class="flex flex-row items-center gap-4 justify-between md:justify-start"><label for="CheckOut">CheckOut</label><input type="text" class="p-inputtext p-component p-inputmask w-32 text-gray-700 font-medium focus:outline-none focus:ring-1 border rounded-lg px-3 py-2 focus:ring-emerald-500" placeholder="dd/mm/yyyy" data-pc-name="pcinputtext" data-pc-extend="inputtext" data-pc-section="root" value=""></div>"
`;
